src/CFG-based-system.typ

```diff
@@ -1 +1,141 @@
-= CFG Based System
\ No newline at end of file
+#import "./proof-tree.typ": *
+#import "./vars.typ": *
+= CFG Based System
+
+Basic rules for Uniqueness and Borrowing won't change here,
+but adapted to fit the CFG representation of Kotlin.
+
+The key gap between the CFG version and the type rules is at the function call.
+Type checking at function calls need adjustment so that it can be done in control flow order.
+The support for nested function call is alse missing is previous typing rules.
+
+== Formalization
+#grid(
+  columns: (auto, auto),
+  column-gutter: 2em,
+  row-gutter: 1em,
+  [*CFG nodes not covered in Grammar*],[*Context*],
+  frame-box(
+    $
+      af &::= unique | shared \
+      beta &::= dot | borrowed \
+      p &::= x | p.f \
+      "TA" &::= "$"i := p, "TA for temp assignment" \
+      "F" &::= f("$"1: af beta, ..., "$"n: af beta): af \
+      "Join" &::= j \
+    $
+  ),
+  frame-box(
+    $
+      alpha &::= unique | shared | top \
+      beta &::= dot | borrowed \
+      Delta &::= dot | p : alpha beta, Delta \
+      Gamma &::= dot | i : Delta, Gamma
+    $
+  )
+)
+
+
+We follow the original lattice for control flow analysis.
+#figure(image(width: 25%, "./img/lattice.svg"), caption: [Lattice in Control Flow Analysis])<CFA-lattice>
+
+In Kotlin CFG, $"$"i = "eval" e$, is a common pattern that appears everywhere,
+we distinguish this case with assignment statement and call this temp assignment.
+We introduce an extra context $Gamma$ for CFA to track aliasing happening at temp assignment.
+Notice that $Gamma$ is a list of aliasing information, as we might have multiple aliasing possibilities.
+
+Rule for CFA are as follows on different nodes in CFG that does not appear in the grammar.
+#display-rules(row-size: 1,
+prooftree(
+  axiom($Gamma'= i: (p, Delta(p), dot), Gamma$),
```

Comment:
Can it ever happen that `i` is associated with `p, αβ` and `αβ != Δ(p)`?  If not, why store both?

Reply:
The only exception is that `$i` is not an aliasing of a path, but only takes an `αβ` information, for example:
```kt
fun unique g(x)
fun f(x: unique)
fun use_f(x) {
  f(g(x)) // $1 = g(x) {$1: [_: unique]}
}
```

Reply:
I see; should it perhaps be either a path or lattice element, then?